beluga/include/beluga/views/random_intersperse.hpp

```diff
@@ -0,0 +1,185 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_VIEWS_RANDOM_INTERSPERSE_HPP
+#define BELUGA_VIEWS_RANDOM_INTERSPERSE_HPP
+
+#include <functional>
+#include <optional>
+#include <random>
+
+#include <range/v3/functional/bind_back.hpp>
+#include <range/v3/utility/random.hpp>
+#include <range/v3/view/adaptor.hpp>
+
+/**
+ * \file
+ * \brief Implementation of a random_intersperse range adaptor object.
+ */
+
+namespace beluga::views {
+
+namespace detail {
+
+/// Implementation of the random_intersperse view as a view adaptor.
+/**
+ * \tparam Range A [forward range](https://en.cppreference.com/w/cpp/ranges/forward_range).
+ * \tparam Fn A callable type which takes no arguments and returns values to be inserted.
+ * \tparam URNG A random number generator that satisfies the
+ *  [UniformRandomBitGenerator](https://en.cppreference.com/w/cpp/named_req/UniformRandomBitGenerator)
+ *  requirements.
+ */
+template <class Range, class Fn, class URNG = typename ranges::detail::default_random_engine>
+struct random_intersperse_view
+    : public ranges::view_adaptor<
+          random_intersperse_view<Range, Fn, URNG>,
+          Range,
+          // The cardinality value can be infinite, unknown, finite or a specific number if known.
+          // If the adapted range cardinality is finite (or a specific number) then we know the
+          // resulting view is finite. Else, we propagate the infinite or unknown specification.
+          // Care must be taken when the adapted range is finite but the intersperse probability
+          // is 1.0, leading to an infinite range in practice.
```

Comment:
@nahueespinosa is it accounted for? It doesn't seem like it. Shouldn't the view constructor `throw` or `assert` in that case?

Reply:
I didn't want to limit the probability values, I use 1.0 in tests too. So I simply specified an `unkown` cardinality here https://github.com/Ekumen-OS/beluga/pull/278/commits/63dffa9881a85f94786717f73509a35b5c15a302.